data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultCosmosRepositoryOperations.java

```diff
@@ -704,40 +712,116 @@ private boolean isIdPartitionKey(PersistentEntity persistentEntity) {
      * @param persistentEntity the persistent entity
      * @return partition key definition it exists for persistent entity, otherwise empty/null string
      */
-    private String getPartitionKey(PersistentEntity persistentEntity) {
+    @NonNull
+    private String getPartitionKeyDefinition(PersistentEntity persistentEntity) {
         CosmosDatabaseConfiguration.CosmosContainerSettings cosmosContainerSettings = cosmosContainerSettingsMap.get(persistentEntity.getPersistedName());
         if (cosmosContainerSettings == null) {
-            return null;
+            return Constants.NO_PARTITION_KEY;
         }
         return CosmosDatabaseInitializer.getPartitionKey(persistentEntity, cosmosContainerSettings);
     }
 
     /**
-     * Gets partition key value for a document. Partition keys can be only string or number values.
+     * Gets partition key for a document. Partition keys can be only string or number values.
      * TODO: Later deal with nested paths when we support it.
      *
      * @param persistentEntity the persistent entity
      * @param item item from the Cosmos Db
-     * @return partition key value, if partition key defined and value set
+     * @return partition key, if partition key defined and value set otherwise null
      */
-    private Object getPartitionKeyValue(RuntimePersistentEntity<?> persistentEntity, ObjectNode item) {
-        String partitionKey = getPartitionKey(persistentEntity);
-        if (StringUtils.isEmpty(partitionKey)) {
+    @Nullable
+    private PartitionKey getPartitionKey(RuntimePersistentEntity<?> persistentEntity, ObjectNode item) {
+        String partitionKeyDefinition = getPartitionKeyDefinition(persistentEntity);
+        if (partitionKeyDefinition.startsWith("/")) {
+            partitionKeyDefinition = partitionKeyDefinition.substring(1);
+        }
+        com.fasterxml.jackson.databind.JsonNode jsonNode = item.get(partitionKeyDefinition);
+        if (jsonNode == null) {
             return null;
         }
-        if (partitionKey.startsWith("/")) {
-            partitionKey = partitionKey.substring(1);
+        Object value;
+        if (jsonNode.isNumber()) {
+            value = jsonNode.numberValue();
+        } else if (jsonNode.isBoolean()) {
+            value = jsonNode.booleanValue();
+        } else {
+            value = jsonNode.textValue();
         }
-        com.fasterxml.jackson.databind.JsonNode jsonNode = item.get(partitionKey);
-        if (jsonNode == null) {
-            return  null;
+        return new PartitionKey(value);
+    }
+
+    /**
+     * Gets underlying java class for the {@link DataType}.
+     *
+     * @param dataType the data type
+     * @return java class for the data type
+     */
+    Class<?> getDataTypeClass(DataType dataType) {
```

Comment:
Do we have tests for those projections?

Reply:
There was test for count (LONG) and I added two more tests for STRING and DATE. Not sure what arrays types are used for, I tried but result type resolves to string instead of string[] so I changed this method to return Object type as default so users can cast if needed for specific types. Simple types work fine.